core/include/vecmem/containers/array.hpp

```diff
@@ -0,0 +1,154 @@
+   ///
+   /// However, to be able to use it as a drop-in replacement for @c std::array,
+   /// it does provide an optional second template argument, which can be used
+   /// to set the fixed size of the array at compile time.
+   ///
+   template< typename T, std::size_t N = 0 >
+   class array {
+
+   public:
+      /// @name Type definitions, mimicking @c std::array
+      /// @{
+
+      /// Type of the array elements
+      typedef T                 value_type;
+      /// Size type for the array
+      typedef std::size_t       size_type;
+      /// Pointer difference type
+      typedef std::ptrdiff_t    difference_type;
+
+      /// Value reference type
+      typedef value_type&       reference_type;
+      /// Constant value reference type
+      typedef const value_type& const_reference_type;
+
+      /// Value pointer type
+      typedef value_type*       pointer_type;
+      /// Constant value pointer type
+      typedef const value_type* const_pointer_type;
+
+      /// Forward iterator type
+      typedef typename std::array< value_type, N >::iterator
+         iterator;
+      /// Constant forward iterator type
+      typedef typename std::array< value_type, N >::const_iterator
+         const_iterator;
+
+      /// Reverse iterator type
+      typedef typename std::array< value_type, N >::reverse_iterator
+         reverse_irerator;
+      /// Constant reverse iterator type
+      typedef typename std::array< value_type, N >::const_reverse_iterator
+         const_reverse_irerator;
+
+      /// @}
+
+      /// Struct used for deleting the allocated memory block
+      class deleter {
+
+      public:
+         /// Constructor
+         deleter( std::size_t bytes, memory_resource& resource );
+
+         /// Copy constructor
+         deleter( const deleter& ) = default;
+         /// Move constructor
+         deleter( deleter&& ) = default;
+         /// Copy assignment operator
+         deleter& operator=( const deleter& ) = default;
+         /// Move assignment operator
+         deleter& operator=( deleter&& ) = default;
+
+         /// Operator performing the deletion of the object.
+         void operator()( void* ptr );
+
+      private:
+         /// The number of bytes allocated for the memory block
+         std::size_t m_bytes;
+         /// The memory resource used for deleting the memory block
+         memory_resource* m_resource;
+
+      }; // struct deleter
+
+      /// Constructor with a memory resource to use
+      ///
+      /// Can only be used if the user chose a non-default (non-zero) value
+      /// for the size template parameter.
+      ///
+      array( memory_resource& resource );
+
+      /// Constructor with a size and a memory resource to use
+      ///
+      /// Can only be used if the user uses the default (zero) value for the
+      /// size template parameter.
+      ///
+      array( std::size_t size, memory_resource& resource );
```

Comment:
Maybe SFINAE it out then in case `N!=0`? Also, I'd consider reversing the argument order.

Reply:
I mean, the `static_assert` does almost the same tbf

Reply:
My first thought was also SFINAE. :smile: But...

```
[ 57%] Building CXX object tests/core/CMakeFiles/vecmem_test_core_arrays.dir/test_core_arrays.cpp.o
In file included from /home/krasznaa/ATLAS/projects/vecmem/vecmem/tests/core/test_core_arrays.cpp:9:
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/array.hpp:139:68: error: cannot add a default template argument to the definition of a member of a class template
   template< std::enable_if_t< is_different< N, 0 >::value, bool > = true >
                                                                   ^ ~~~~
/home/krasznaa/ATLAS/projects/vecmem/vecmem/core/include/vecmem/containers/array.hpp:148:64: error: cannot add a default template argument to the definition of a member of a class template
   template< std::enable_if_t< is_equal< N, 0 >::value, bool > = true >
                                                               ^ ~~~~
```

Apparently you can't just provide a single default template argument for SFINAE to work. There would have to be a specific template parameter for the constructor as well.

So after some fumbling I went with `static_assert(...)`...

Reply:
Yeah I think it would have to be something like 

```cpp
template <typename _N = N, std::enable_if_t< _N ==0 >::value, bool > = true
array(std::size_t, memory_resource& resource) {
  // ...
}
```

to lift it into a constructor template. (Pseudo code, this won't compile). In the end it's not really important anyway.